File: labs/04-tool-calling/node/main.ts
import { OpenAIClient, ChatMessage } from "../../../shared/config/openai_client";

const TOOLS = [
  {
    type: "function",
    function: {
      name: "get_weather",
      description: "Get current weather for a city.",
      parameters: {
        type: "object",
        properties: {
          city: { type: "string", description: "City name" },
        },
        required: ["city"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "search_notes",
      description: "Search project notes by keyword.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "Search keyword" },
        },
        required: ["query"],
      },
    },
  },
];

const NOTES = [
  {
    title: "MCP design",
    content: "MCP uses Resources, Tools, Prompts, and Sampling primitives.",
  },
  {
    title: "LSP basics",
    content: "LSP speaks JSON-RPC over stdio or sockets.",
  },
  {
    title: "Agent memory",
    content: "Short-term memory lives in the context window; long-term in vectors.",
  },
];

function getWeather(city: string): string {
  return JSON.stringify({ city, temperature_c: 22, condition: "sunny" });
}

function searchNotes(query: string): string {
  const results = NOTES.filter(
    (n) =>
      n.title.toLowerCase().includes(query.toLowerCase()) ||
      n.content.toLowerCase().includes(query.toLowerCase())
  );
  return JSON.stringify(results);
}

const TOOL_FUNCTIONS: Record<string, (args: any) => string> = {
  get_weather: getWeather,
  search_notes: searchNotes,
};

interface ToolCall {
  id: string;
  function: {
    name: string;
    arguments: string;
  };
}

async function runToolAgent(
  client: OpenAIClient,
  userMessage: string,
  maxIterations = 5
): Promise<string> {
  const messages: ChatMessage[] = [{ role: "user", content: userMessage }];

  for (let i = 0; i < maxIterations; i++) {
    const response = await client.chatCompletion({
      messages,
      tools: TOOLS as any,
      tool_choice: "auto",
      temperature: 0.0,
      max_tokens: 300,
    });
    const message = response.choices[0].message;
    messages.push(message);

    const toolCalls = (message.tool_calls as ToolCall[] | undefined) || [];
    if (toolCalls.length === 0) {
      return message.content;
    }

    for (const toolCall of toolCalls) {
      const name = toolCall.function.name;
      const args = JSON.parse(toolCall.function.arguments);
      console.log(`[Tool call] ${name}(${JSON.stringify(args)})`);
      const func = TOOL_FUNCTIONS[name];
      const result = func
        ? func(args)
        : JSON.stringify({ error: `unknown tool ${name}` });
      messages.push({
        role: "tool",
        tool_call_id: toolCall.id,
        name,
        content: result,
      });
      console.log(`[Tool result] ${result}`);
    }
  }

  return "Reached max iterations.";
}

async function main() {
  const client = new OpenAIClient();
  const question = "What's the weather in Shanghai? Also, find me notes about MCP.";
  console.log("User:", question);
  const answer = await runToolAgent(client, question);
  console.log("\nAssistant:", answer);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: labs/04-tool-calling/python/main.py
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "shared" / "config"))

from openai_client import OpenAIClient, configure_logging


TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": "Get current weather for a city.",
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {"type": "string", "description": "City name"},
                },
                "required": ["city"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search_notes",
            "description": "Search project notes by keyword.",
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {"type": "string", "description": "Search keyword"},
                },
                "required": ["query"],
            },
        },
    },
]


NOTES = [
    {"title": "MCP design", "content": "MCP uses Resources, Tools, Prompts, and Sampling primitives."},
    {"title": "LSP basics", "content": "LSP speaks JSON-RPC over stdio or sockets."},
    {"title": "Agent memory", "content": "Short-term memory lives in the context window; long-term in vectors."},
]


def get_weather(city: str) -> str:
    return json.dumps({"city": city, "temperature_c": 22, "condition": "sunny"})


def search_notes(query: str) -> str:
    results = [n for n in NOTES if query.lower() in n["title"].lower() or query.lower() in n["content"].lower()]
    return json.dumps(results, ensure_ascii=False)


TOOL_FUNCTIONS = {
    "get_weather": get_weather,
    "search_notes": search_notes,
}


def run_tool_agent(client: OpenAIClient, user_message: str, max_iterations: int = 5) -> str:
    messages = [{"role": "user", "content": user_message}]

    for _ in range(max_iterations):
        response = client.chat_completion(
            messages=messages,
            tools=TOOLS,
            tool_choice="auto",
            temperature=0.0,
            max_tokens=300,
        )
        message = response["choices"][0]["message"]
        messages.append(message)

        if not message.get("tool_calls"):
            return message["content"]

        for tool_call in message["tool_calls"]:
            name = tool_call["function"]["name"]
            args = json.loads(tool_call["function"]["arguments"])
            print(f"[Tool call] {name}({args})")
            func = TOOL_FUNCTIONS.get(name)
            result = func(**args) if func else json.dumps({"error": f"unknown tool {name}"})
            messages.append(
                {
                    "role": "tool",
                    "tool_call_id": tool_call["id"],
                    "name": name,
                    "content": result,
                }
            )
            print(f"[Tool result] {result}")

    return "Reached max iterations."


def main():
    configure_logging()
    client = OpenAIClient()
    question = "What's the weather in Shanghai? Also, find me notes about MCP."
    print("User:", question)
    answer = run_tool_agent(client, question)
    print("\nAssistant:", answer)


if __name__ == "__main__":
    main()
